packages/xen-api/src/index.js

```diff
@@ -1004,6 +1008,53 @@ export class Xapi extends EventEmitter {
       delete taskWatchers[ref]
     }
   }
+  async _findCurrentMaster() {
+    debug('%s: will try to find a new master', this._humanId)
+    const hosts = await this.getAllRecords('host')
+    const hostAddresses = this.fallBackAddresses.concat(hosts.map(({ address }) => address))
+
+    // there can be a master only if there is more than 1 host
+    if (hostAddresses.length < 2) {
+      return Promise.resolve()
+    }
+    // if we were already connected
+    for (const hostAdress of hostAddresses) {
+      // not current master send him a simple query and
+      if (hostAdress !== this.url.hostname) {
+        continue
+      }
+      // @todo handle _reverseHostIpAddresses
+      const url = { ...this.url, hostname: hostAdress }
+      // instantiate a transport to this url
+      const transport = autoTransport({
+        secureOptions: {
+          minVersion: 'TLSv1',
+          rejectUnauthorized: !this._allowUnauthorized,
+        },
+        url,
+        httpProxy: this._httpProxy,
+      })
```

Comment:
Why not use `_setUrl`?

Reply:
there may be multiple calls to this in parallel, I'm trying to update the url only on success. That way even if I call multiple time setUrl, it will be with the same parameters ( given that the new master does not fail during the interval )